### PR TITLE
Separate serial consistency configuration from remoteRouting functionality - Issue #633

### LINKED ISSUE #633: Separate serial consistency configuration from remoteRouting functionality
Today remoteRouting do two things:

- it decides if look requests are routed to the local DC or to the DC that the node you try to lock belongs to.
- it selects the serial consistency so the we use LOCAL_SERIAL if we have remoteRouting true and SERIAL is its false.

The idea is to leave the remoteRouting option as it is and add a new option that can override the serial consistency elected by remoteRouting, so it might have three settings (DEFAULT/SERIAL/LOCAL_SERIAL). and default give whatever remoteRouting selected, and the other two override that just force SERIAL or LOCAL_SERIAL regardless of the remoteRouting setting.

### PR DESCRIPTION
**Pull request by VictorCavichioli**
To decouple the consistency definition in lightweight transactions, introduce a new property configurable by ecc.yml that allows consistency to be overridden by the values DEFAULT/SERIAL/LOCAL.

Closes #633

### CODE REVIEW

File: application/src/main/resources/ecc.yml [hunk 1]
```diff
@@ -53,10 +53,17 @@ connection:
     ## Allow routing requests directly to a remote datacenter.
     ## This allows locks for other datacenters to be taken in that datacenter instead of via the local datacenter.
     ## If clients are prevented from connecting directly to Cassandra nodes in other sites this is not possible.
-    ## If remote routing is disabled its not possible to use LOCAL_SERIAL consistency for the locking,
+    ## If remote routing is disabled and consistency serial uses the DEFAULT mode its not possible to use LOCAL_SERIAL consistency for the locking,
     ## instead SERIAL consistency will be used for those request.
     ##
     remoteRouting: true
+    ##
```
Comment (masokol) [inline application/src/main/resources/ecc.yml:60]: Try to follow the same format as above. Something like:
Allow to override consistency level for LWT (lightweight transactions).
Possible values are:
  "DEFAULT" - Use consistency level based on remoteRouting.
  "SERIAL" - Use SERIAL consistency for LWT regardless of remoteRouting.
  "LOCAL_SERIAL" - Use LOCAL_SERIAL consistency for LWT regardless of remoteRouting.

File: application/src/main/resources/ecc.yml [hunk 2]
```diff
@@ -53,10 +53,17 @@ connection:
     ## Allow routing requests directly to a remote datacenter.
     ## This allows locks for other datacenters to be taken in that datacenter instead of via the local datacenter.
     ## If clients are prevented from connecting directly to Cassandra nodes in other sites this is not possible.
-    ## If remote routing is disabled its not possible to use LOCAL_SERIAL consistency for the locking,
+    ## If remote routing is disabled and consistency serial uses the DEFAULT mode its not possible to use LOCAL_SERIAL consistency for the locking,
```
Comment (masokol) [inline application/src/main/resources/ecc.yml:56]: Remove "its not possible to use LOCAL_SERIAL consistency for the locking,", this is not true anymore since you can override it with consistencySerial

File: application/src/main/resources/ecc.yml [hunk 3]
```diff
@@ -53,10 +53,17 @@ connection:
...
     ## instead SERIAL consistency will be used for those request.
     ##
     remoteRouting: true
+    ##
+    # consistencySerial value can be DEFAULT, which means that the serial consistency for lightweight transactions is based
+    # on the remoteRouting configurations. It is possible to define as LOCAL or SERIAL as well, the first one defines serial consistency
+    # as LOCAL_SERIAL which requires a quorum of replicas in the local DC, also SERIAL defines a serial consistency of SERIAL
```
Comment (masokol) [inline application/src/main/resources/ecc.yml:63]: I don't think we need to explain what LOCAL_SERIAL and SERIAL is. Instead, we should explain the implications, i.e if you use remoteRouting: false and LOCAL_SERIAL here then all locks will be taken locally in DC. I.e There's a risk that multiple nodes in different datacenters will be able to lock the same nodes -> multiple repairs on the same range/node at the same time.

**Review by masokol — COMMENTED** [review]
Is this an enhancement or bug? If it's a bug then we should aim for lowest supported version and merge forward.

File: application/src/main/resources/ecc.yml (hunk 1 above)
Comment (VictorCavichioli) [reply application/src/main/resources/ecc.yml:59]: Done

**Comment by VictorCavichioli** [comment]
> Is this an enhancement or bug? If it's a bug then we should aim for lowest supported version and merge forward.

It's an enhancement but as we discusses we will create another PR to ecchronos 4.x and merge first

File: application/src/main/resources/ecc.yml (hunk 3 above)
Comment (VictorCavichioli) [reply application/src/main/resources/ecc.yml]: Done

File: application/src/main/resources/ecc.yml [hunk 4]
```diff
@@ -53,10 +53,20 @@ connection:
     ## Allow routing requests directly to a remote datacenter.
     ## This allows locks for other datacenters to be taken in that datacenter instead of via the local datacenter.
     ## If clients are prevented from connecting directly to Cassandra nodes in other sites this is not possible.
-    ## If remote routing is disabled its not possible to use LOCAL_SERIAL consistency for the locking,
-    ## instead SERIAL consistency will be used for those request.
+    ## If remote routing is disabled, instead SERIAL consistency will be used for those request.
     ##
     remoteRouting: true
+    ##
+    # # Allow to override consistency level for LWT (lightweight transactions). Possible values are:
```
Comment (masokol) [inline application/src/main/resources/ecc.yml:60]: Is there a reason for having spaces between "#"? If not then remove them.

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/AbstractCassandraTest.java [hunk 5]
```diff
@@ -77,6 +77,11 @@ public boolean getRemoteRouting()
             {
                 return true;
             }
+
+            @Override
+            public String getSerialConsistency(){
```
Comment (masokol) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/AbstractCassandraTest.java:82]: Incorrect formatting {} should be placed on newlines.

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java [hunk 6]
```diff
@@ -375,13 +380,103 @@ public boolean getRemoteRouting()
                             {
                                 return true;
                             }
+
+                            @Override
+                            public String getSerialConsistency(){
```
Comment (masokol) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java:385]: Incorrect formatting {} should be placed on newlines.

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java [hunk 7]
```diff
@@ -375,13 +380,103 @@ public boolean getRemoteRouting()
...
+                            }
                         })
                         .withHostStates(hostStates)
                         .withStatementDecorator(s -> s)
                         .withKeyspaceName(myKeyspaceName)
                         .build());
     }
 
+    @Test
+    public void testRemoteRoutingTrueWithDefaultSerialConsistency() {
```
Comment (masokol) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java:396]: Incorrect formatting {} should be placed on newlines.

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java [hunk 8]
```diff
@@ -375,13 +380,103 @@ public boolean getRemoteRouting()
...
+                .withStatementDecorator(s -> s)
+                .withKeyspaceName(myKeyspaceName)
+                .build();
+
+        
+        assertEquals(ConsistencyLevel.LOCAL_SERIAL, myLockFactory.getSerialConsistencyLevel());
+    }
+
+    @Test
+    public void testRemoteRoutingFalseWithDefaultSerialConsistency() {
```
Comment (masokol) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java:410]: Incorrect formatting {} should be placed on newlines.

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java [hunk 9]
```diff
@@ -375,13 +380,103 @@ public boolean getRemoteRouting()
...
+                .withHostStates(hostStates)
+                .withStatementDecorator(s -> s)
+                .withKeyspaceName(myKeyspaceName)
+                .build();
+
+        assertEquals(ConsistencyLevel.SERIAL, myLockFactory.getSerialConsistencyLevel());
+    }
+
+    @Test
+    public void testLocalSerialConsistency(){
```
Comment (masokol) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java:435]: Incorrect formatting {} should be placed on newlines.

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java [hunk 10]
```diff
@@ -375,13 +380,103 @@ public boolean getRemoteRouting()
...
+                .withStatementDecorator(s -> s)
+                .withKeyspaceName(myKeyspaceName)
+                .build();
+
+        assertEquals(ConsistencyLevel.LOCAL_SERIAL, myLockFactory.getSerialConsistencyLevel());
+
+    }
+
+    @Test
+    public void testSerialConsistency(){
```
Comment (masokol) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java:459]: Incorrect formatting {} should be placed on newlines.

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/repair/TestDefaultRepairConfigurationProvider.java [hunk 11]
```diff
@@ -101,6 +101,11 @@ public boolean getRemoteRouting()
             {
                 return true;
             }
+
+            @Override
+            public String getSerialConsistency(){
```
Comment (masokol) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/repair/TestDefaultRepairConfigurationProvider.java:106]: Incorrect formatting {} should be placed on newlines.

File: pmd-rules.xml [hunk 12]
```diff
@@ -30,6 +30,10 @@
         <exclude name="DataClass" />
         <exclude name="ExcessiveImports" />
         <exclude name="AvoidUncheckedExceptionsInSignatures" />
+        <exclude name="GodClass" />
```
Comment (masokol) [inline pmd-rules.xml:33]: What, did you agree to modify these or is it just for testing?

File: application/src/main/resources/ecc.yml (hunk 4 above)
Comment (VictorCavichioli) [reply application/src/main/resources/ecc.yml]: Done

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/AbstractCassandraTest.java (hunk 5 above)
Comment (VictorCavichioli) [reply core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/AbstractCassandraTest.java]: Done

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java (hunk 6 above)
Comment (VictorCavichioli) [reply core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java]: Done

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java (hunk 7 above)
Comment (VictorCavichioli) [reply core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java]: Done

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java (hunk 10 above)
Comment (VictorCavichioli) [reply core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java]: Done

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/repair/TestDefaultRepairConfigurationProvider.java (hunk 11 above)
Comment (VictorCavichioli) [reply core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/repair/TestDefaultRepairConfigurationProvider.java]: Done

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/NativeConnection.java [hunk 13]
```diff
@@ -32,6 +32,7 @@ public class NativeConnection extends Connection<NativeConnectionProvider>
 
     private Class<? extends StatementDecorator> myDecoratorClass = NoopStatementDecorator.class;
     private boolean myRemoteRouting = true;
+    private String myConsistencySerial = "DEFAULT";
```
Comment (masokol) [inline application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/NativeConnection.java:35]: Why not an enum?

File: pmd-rules.xml [hunk 14]
```diff
@@ -30,6 +30,10 @@
         <exclude name="DataClass" />
         <exclude name="ExcessiveImports" />
         <exclude name="AvoidUncheckedExceptionsInSignatures" />
+        <exclude name="GodClass" /> <!-- Should consider fixing this -->
```
Comment (masokol) [inline pmd-rules.xml:33]: This xml file is affecting all files in the repo. If you really want to suppress something because of this patch then use annotations "@SuppressWarnings"

File: pmd-rules.xml (hunk 12 above)
Comment (VictorCavichioli) [reply pmd-rules.xml]: As we discussed, for now we will use SuppressWarning annotation to avoid this.

File: pmd-rules.xml (hunk 14 above)
Comment (VictorCavichioli) [reply pmd-rules.xml]: Done

File: application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/NativeConnection.java (hunk 13 above)
Comment (VictorCavichioli) [reply application/src/main/java/com/ericsson/bss/cassandra/ecchronos/application/config/connection/NativeConnection.java]: Done

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/utils/ConsistencyType.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2023 Telefonaktiebolaget LM Ericsson
```
Comment (masokol) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/utils/ConsistencyType.java:2]: it's 2024 now.

File: pmd-rules.xml [hunk 16]
```diff
@@ -49,4 +49,4 @@
             <property name="skipAnnotations" value="true" />
```
Comment (masokol) [inline pmd-rules.xml:49]: Remove this file so it's not part of this PR.

Comment (VictorCavichioli) [reply pmd-rules.xml]: Done

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/utils/ConsistencyType.java (hunk 15 above)
Comment (VictorCavichioli) [reply core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/utils/ConsistencyType.java]: Done

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/CASLockFactory.java [hunk 17]
```diff
@@ -124,12 +129,23 @@ private CASLockFactory(final Builder builder)
 
         mySession = builder.myNativeConnectionProvider.getSession();
         myRemoteRouting = builder.myNativeConnectionProvider.getRemoteRouting();
+        mySerialConsistency = builder.myNativeConnectionProvider.getSerialConsistency();
 
         verifySchemasExists();
 
-        ConsistencyLevel serialConsistencyLevel = myRemoteRouting
+        if (ConsistencyType.DEFAULT.getStringValue().equals(mySerialConsistency))
```
[the remark this answers is not in the document]
Comment (VictorCavichioli) [reply core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/CASLockFactory.java]: Done

File: application/src/test/java/com/ericsson/bss/cassandra/ecchronos/application/config/TestConfig.java [hunk 18]
```diff
@@ -271,6 +273,7 @@ public void testDefault() throws Exception
         assertThat(nativeConnection.getHost()).isEqualTo("localhost");
         assertThat(nativeConnection.getPort()).isEqualTo(9042);
         assertThat(nativeConnection.getRemoteRouting()).isTrue();
+
```
Comment (masokol) [inline application/src/test/java/com/ericsson/bss/cassandra/ecchronos/application/config/TestConfig.java:276]: Remove this newline (l276)

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/CASLockFactory.java [hunk 19]
```diff
@@ -102,6 +104,7 @@ public final class CASLockFactory implements LockFactory, Closeable
     private final StatementDecorator myStatementDecorator;
     private final HostStates myHostStates;
     private final boolean myRemoteRouting;
+    private final ConsistencyType mySerialConsistency;
```
Comment (masokol) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/CASLockFactory.java:107]: This variable is only used in constructor so there's no point in having it.

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/CASLockFactory.java [hunk 20]
```diff
@@ -124,12 +129,23 @@ private CASLockFactory(final Builder builder)
 
         mySession = builder.myNativeConnectionProvider.getSession();
         myRemoteRouting = builder.myNativeConnectionProvider.getRemoteRouting();
+        mySerialConsistency = builder.myConsistencyType;
 
         verifySchemasExists();
 
-        ConsistencyLevel serialConsistencyLevel = myRemoteRouting
+        if (ConsistencyType.DEFAULT.equals(mySerialConsistency))
```
Comment (masokol) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/CASLockFactory.java:136]: use builder.myConsistencyType instead

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/CASLockFactory.java [hunk 21]
```diff
@@ -124,12 +129,23 @@ private CASLockFactory(final Builder builder)
...
+        {
+            serialConsistencyLevel = myRemoteRouting
                 ? ConsistencyLevel.LOCAL_SERIAL
                 : ConsistencyLevel.SERIAL;
+        }
+        else
+        {
+            serialConsistencyLevel = ConsistencyType.LOCAL.equals(mySerialConsistency)
```
Comment (masokol) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/CASLockFactory.java:144]: use builder.myConsistencyType instead

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/CASLockFactory.java [hunk 22]
```diff
@@ -114,6 +117,8 @@ public final class CASLockFactory implements LockFactory, Closeable
     private final PreparedStatement myRemoveLockPriorityStatement;
     private final CASLockFactoryCacheContext myCasLockFactoryCacheContext;
 
+    private final ConsistencyLevel serialConsistencyLevel;
```
Comment (masokol) [inline core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/CASLockFactory.java:120]: mySerialConsistencyLevel

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java
Comment (masokol) [inline core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java:1]: Nit: too many newlines, i don't see any reason why you would want a newline after each line.

File: CHANGES.md [hunk 23]
```diff
@@ -2,10 +2,11 @@
 
 ## Version 5.0.1 (Not yet released)
 
+* Separate serial consistency configuration from remoteRouting functionality - Issue #633
```
Comment (masokol) [inline CHANGES.md:5]: Update the title of the PR to this.

File: core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java
Comment (VictorCavichioli) [reply core/src/test/java/com/ericsson/bss/cassandra/ecchronos/core/TestCASLockFactory.java:1]: Done

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/CASLockFactory.java (hunk 22 above)
Comment (VictorCavichioli) [reply core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/CASLockFactory.java]: Done

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/CASLockFactory.java (hunk 21 above)
Comment (VictorCavichioli) [reply core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/CASLockFactory.java]: Done

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/CASLockFactory.java (hunk 20 above)
Comment (VictorCavichioli) [reply core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/CASLockFactory.java]: Done

File: application/src/test/java/com/ericsson/bss/cassandra/ecchronos/application/config/TestConfig.java (hunk 18 above)
Comment (VictorCavichioli) [reply application/src/test/java/com/ericsson/bss/cassandra/ecchronos/application/config/TestConfig.java]: Done

**Review by jwaeab — APPROVED** [review]
